### PR TITLE
Rewards Greaselion - Twitter

### PR DESCRIPTION
**Pull request by emerick**
Resolves https://github.com/brave/brave-browser/issues/11462
Resolves https://github.com/brave/brave-browser/issues/6762
Resolves https://github.com/brave/brave-browser/issues/5685
Resolves https://github.com/brave/brave-browser/issues/4557
Resolves https://github.com/brave/brave-browser/issues/8265

Associated Greaselion script PR: https://github.com/brave/brave-site-specific-scripts/pull/16

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`, `npm run gn_check`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

**Note**: In order to test this PR, you must manually install the Twitter Greaselion script at the moment. This involves pulling down this branch and running the following commands (the example is Windows specific, it will need to be modified slightly for Mac/Linux):

```
npm run build
cp -R dist/Greaselion.json /c/Users/$USERNAME/AppData/Local/BraveSoftware/Brave-Browser/User\ Data/afalakplffnnnlkncjhbmahjfjhmlkal/1.0.28/1/
cp -R dist/scripts /c/Users/$USERNAME/AppData/Local/BraveSoftware/Brave-Browser/User\ Data/afalakplffnnnlkncjhbmahjfjhmlkal/1.0.28/1/
```

#### Standard URL

- Clean profile
- Run browser
- Install Twitter Greaselion script as explained above
- Restart browser
- Visit a Twitter URL (e.g., https://twitter.com/emerick)
- Verify that correct publisher information appears in Rewards panel
- Verify that tipping works as expected
- Verify that verified checkmark appears in Rewards icon for verified publishers

#### Excluded URL

- Clean profile
- Run browser
- Install Twitter Greaselion script as explained above
- Restart browser
- Visit an excluded Twitter URL (e.g., https://twitter.com/settings)
- Verify that correct publisher information appears in Rewards panel (twitter.com)
- Verify that tipping works as expected

#### Activity

- Clean profile
- Run browser
- Install Twitter Greaselion script as explained above
- Restart browser
- Visit a Twitter URL (e.g., https://twitter.com/emerick)
- Verify that `activity_info` database table is updated as expected
- Verify that auto-contribute table is updated to reflect new activity

#### History

- Navigate to various Twitter URLs in the same tab
- Verify that appropriate publisher information appears in the panel
- Verify that back/forward navigation updates the panel with correct publisher info

#### Tab Activation

- Navigate to various Twitter URLs in multiple tabs
- Verify that appropriate publisher information appears in the panel
- Verify that switching between tabs updates the panel with correct publisher info

#### Tipping non-Twitter Publishers

- Verify that tipping works as expected on non-Twitter publishers (we only made changes to Twitter and only on desktop)
- Verify that the panel looks as expected on non-Twitter publishers

### Mobile

No changes were made to mobile platforms, so verify that Twitter navigation / tipping / Rewards panel continue to work as expected on those platforms.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: browser/extensions/api/brave_rewards_api.cc
Comment (NejcZdovc) [inline browser/extensions/api/brave_rewards_api.cc:329]: can we do something like this 

```
if (result == ledger::type::Result::LEDGER_OK) {
  ShowTipDialog();
  Release();
 return;
}
```

and then logic in this if after that. This way we don't have this big if

File: vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.cc
Comment (NejcZdovc) [inline vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.cc:921]: can you please add unit tests for this function

File: browser/extensions/api/brave_rewards_api.cc
Comment (emerick) [reply browser/extensions/api/brave_rewards_api.cc:329]: Fixed.

File: vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.cc
Comment (emerick) [reply vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.cc:921]: Good idea, added.

File: components/brave_rewards/resources/extension/brave_rewards/background/publisherVisit.ts
Comment (zenparsing) [inline components/brave_rewards/resources/extension/brave_rewards/background/publisherVisit.ts:98]: It looks like we're sending both errors and valid responses through the "response" field into the content script. Does the content script know how to tell the difference? Perhaps the message to the content script should include both "response" and "error" fields?

Comment (emerick) [reply components/brave_rewards/resources/extension/brave_rewards/background/publisherVisit.ts]: Great point, fixed.

File: components/brave_rewards/resources/extension/brave_rewards/background/publisherVisit.ts
Comment (emerick) [reply components/brave_rewards/resources/extension/brave_rewards/background/publisherVisit.ts]: I had `Request` and `Response` appended to all message types for a while when writing the YouTube script. I feel like it didn't really help with clarity and just made things wordier. Once the background script began making requests, it only added to the confusion (at least for me!) I'm going to try to stick with this nomenclature for consistency right now.

**Review by zenparsing — APPROVED** [review]

File: browser/extensions/api/brave_rewards_api.cc
Comment (NejcZdovc) [inline browser/extensions/api/brave_rewards_api.cc:372]: is it possible that this is null? if so please add null check

File: components/brave_rewards/resources/tip/components/app.tsx
Comment (NejcZdovc) [inline components/brave_rewards/resources/tip/components/app.tsx:68]: nit: can this be moved back where it was so that change is removed?

File: vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.cc
Comment (NejcZdovc) [inline vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.cc:928]:
```suggestion
  if (comment == args.end() || name == args.end() || hashtag == args.end()) {
    return "";
  }
```

File: browser/extensions/api/brave_rewards_api.cc
Comment (emerick) [reply browser/extensions/api/brave_rewards_api.cc:372]: Fixed.

File: components/brave_rewards/resources/tip/components/app.tsx
Comment (emerick) [reply components/brave_rewards/resources/tip/components/app.tsx]: Fixed.

File: vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.cc
Comment (emerick) [reply vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.cc]: Fixed.

**Review by NejcZdovc — CHANGES_REQUESTED** [review]
1) when you are on twitter.com we should display just domain
image

2) publisher has different images (panel vs tipping banner)
image
image

3) inline tip is missing when you are on a post 
https://twitter.com/queentatiana/status/1228717509337079808
image

4) panel is not showing publisher info. Visit https://twitter.com/lukemulks/status/1293113074317049856
image

5) wrong publisher in the panel when you visit like page https://twitter.com/brave/likes, https://twitter.com/brave/topics
image

6) no publisher info when you visit with_replies https://twitter.com/brave/with_replies, https://twitter.com/brave/media
image

7) message page should just show twitter.com
image

**Comment by emerick** [comment]
@NejcZdovc Resolved all of the issues you mentioned, including displaying the banner logo in the Rewards panel when available. Should be ready for review again.

**Review by NejcZdovc — CHANGES_REQUESTED** [review]
1) verified publisher doesn't have profile image https://twitter.com/lukemulks/status/1293113074317049856
image


2) should show publisher info https://twitter.com/brave/likes
current:
image
expected:
image

3) inline tip is not showing correct publisher. It should show TNW Events, but it shows Geertje as it was cited by TNW
https://twitter.com/tnwevents/status/1311659702598668288
current:
image
expected:
image

**Review by tmancey — APPROVED** [review]
LGTM++

**Review by NejcZdovc — APPROVED** [review]

**Comment by emerick** [comment]
All CI platforms passed, waiting for final approval on https://github.com/brave/brave-site-specific-scripts/pull/16 before merging.

**Comment by johntringham** [comment]
I believe this change has caused an issue where the embedded tip button can't be turned off on for twitter, github or reddit. There was a test to check whether disabling Rewards also disabled the Tip button, but this PR includes a change that disables it [in this commit](https://github.com/brave/brave-core/commit/1903d54f67fb2cc30ecebda7d48a5a1f3f972963#diff-6a031ef8a596db9fc9456b9b04a9c9f8d80ae82d8f381132401607c92feb26b2). 

I might be wrong about this (first time looking at the codebase) but it looks like this PR rewrites the way tips are created/rendered, but doesn't include any reference to "props.rewardsData.inlineTip" preference, and so will be on by default no matter what the changes are.

**Comment by emerick** [comment]
@johntringham After you disable the tip button, it requires you to restart the browser (we're working on removing that limitation). If you do that, is the button disabled for you?

**Comment by johntringham** [comment]
Yep, I've tried that (and several other things) - the Tip icon still appears. It looks like this bug is there for Twitter, Github and Reddit. 

This has been raised in several places over the last couple of months:
https://community.brave.com/t/no-way-of-removing-tip-button-from-twitter/188046/4
https://community.brave.com/t/twitter-showing-bat-tip-button-when-turned-off-causing-glitching/168542
https://github.com/brave/brave-browser/issues/13223
https://github.com/brave/brave-browser/issues/13222
https://github.com/brave/brave-browser/issues/13157
https://github.com/brave/brave-browser/issues/12870

**Comment by emerick** [comment]
@johntringham No worries at all, glad it's working.

**Pull request merged**
